### PR TITLE
fix sort performance issues for dropdowns

### PR DESCRIPTION
**Pull request by MaximilianKrambach**
This is an alternative approach to #4398, doing the sorting at a more deeper level (close to the data store).

Maybe we can build on either of these two PR (both exclude each other) to get a fast sorting of dropdown lists etc.

### CODE REVIEW

**Comment by MaximilianKrambach** [comment]
This doesn't work, as the background `ViewModelStore` is an object and not an array. The only acceptable part here is the generalization of the sorting, but not the place(s) where the sorting occurs.

(Partial?) sorting should be done in `base-repository:getViewModelListObservable()` if (a) `this.sortFunction` is defined and (b) after a delay of `auditTime( >= 1)`.

**Comment by MaximilianKrambach** [comment]
@emanuelschuetze Finn and I now have a (hopefully) working and performant version; 
the category sorting and the  in the 'sort motions in category' is fixed here, too.

File: client/src/app/core/repositories/motions/motion-repository.service.ts [hunk 1]
```diff
@@ -801,4 +802,25 @@ export class MotionRepositoryService extends BaseAgendaContentObjectRepository<V
             return value;
         }
     }
+
+    /**
+     * Triggers an update for the sort function responsible for the default sorting of data items
+     */
+    public setConfigSortFn(): void {
```
Comment (tsiegleauq) [inline client/src/app/core/repositories/motions/motion-repository.service.ts]: Motion default sorting is missing.
The config has the property:
"Sort motions by" which can either be "Identifier" or "Call list"

Should probably look like in the category repo.

**Comment by MaximilianKrambach** [comment]
Everything seems ready now, but I'd rather have another review by @FinnStutzenstein to double check we don't mess up some unsorted/sorted getViewModelList() instances

File: client/src/app/core/repositories/base-repository.ts [hunk 2]
```diff
@@ -112,7 +147,7 @@ export abstract class BaseRepository<V extends BaseViewModel, M extends BaseMode
                     const viewModel = this.viewModelStoreService.get(model.collectionString, model.id);
 
                     // if an domain object we need was added or changed, update viewModelStore
-                    this.getViewModelList().forEach(ownViewModel => {
+                    Object.values(this.viewModelStore).forEach(ownViewModel => {
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/base-repository.ts]: undo

File: client/src/app/core/repositories/base-repository.ts [hunk 3]
```diff
@@ -205,10 +261,10 @@ export abstract class BaseRepository<V extends BaseViewModel, M extends BaseMode
     }
 
     /**
-     * update the observable of the list
+     * update the observable of the list. Also updates the sorting of the view model list.
      */
     protected updateViewModelListObservable(): void {
-        this.viewModelListSubject.next(this.getViewModelList());
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/base-repository.ts]: undo

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 4]
```diff
@@ -460,25 +460,22 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit,
         this.getMotionByUrl();
         this.setSurroundingMotions();
 
-        // TODO: Changed to un-sort, since it's a really heavy operation
-        this.userRepo.getViewModelListObservable().subscribe(unsortedUsers => {
-            this.submitterObserver.next(unsortedUsers);
-            this.supporterObserver.next(unsortedUsers);
+        // Initial Filling of the Subjects
+        const initialUsers = this.userRepo.getViewModelList();
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: use sorted list

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 5]
```diff
@@ -460,25 +460,22 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit,
...
+        this.submitterObserver = new BehaviorSubject(initialUsers);
+        this.supporterObserver = new BehaviorSubject(initialUsers);
+        this.categoryObserver = new BehaviorSubject(this.categoryRepo.getViewModelList());
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: use sorted

**Comment by MaximilianKrambach** [comment]
Should be in an  okayish state now.

Attention for testing:
- initial sorting in list views
- loading of dropdown-heavy detail views (e.g. a motion detail edit that loads lots of users for submitters/Supporters)
- language change should change language-specific sorting (or at least not crash/stall the sorting)

**Comment by MaximilianKrambach** [comment]
(rebase just resolved the new conflicts). I'm more or less waiting for some performance testing 'in the wild' with real user data; else, this is mergable as I see it

**Comment by emanuelschuetze** [comment]
Still open:
* category sort
* mediafile sort
* if possible in this pr: parent item field should show agenda item (by using getTitle() for topics)

**Comment by MaximilianKrambach** [comment]
> * 
> * if possible in this pr: parent item field should show agenda item (by using getTitle() for topics)

This is not part of this PR, it got broken elsewhere (`search-value-selector` or `getTitle()` in ViewItem or somewhere further down the stack)

File: client/src/app/site/projector/components/projector-list/projector-list.component.html [hunk 6]
```diff
@@ -19,7 +19,7 @@ <h2 translate>Projectors</h2>
         Reference projector for current list of speakers:
     </span>
     <mat-form-field>
-        <mat-select [disabled]="!!editId" [value]="projectors[0].reference_projector_id" (selectionChange)="onSelectReferenceProjector($event)">
+        <mat-select [disabled]="!!editId" [value]="projectors.length ? projectors[0].reference_projector_id : null" (selectionChange)="onSelectReferenceProjector($event)">
```
Comment (MaximilianKrambach) [inline client/src/app/site/projector/components/projector-list/projector-list.component.html]: @FinnStutzenstein This was one of the two travis testing fails that popped up. Please confirm that m change doesn't destroy anything

**Comment by MaximilianKrambach** [comment]
I fixed some copy-paste errors (that caused nothing to be sorted at all for some views), and added a mediafile default sorting (by title). Also, initial sorting on a list view is done now, even if the client has never opened the view before

File: client/src/app/site/projector/components/projector-list/projector-list.component.html (hunk 6 above)
Comment (FinnStutzenstein) [reply client/src/app/site/projector/components/projector-list/projector-list.component.html]: Should be ok.

**Comment by MaximilianKrambach** [comment]
@emanuelschuetze I fixed the first error, and could not reproduce the second one (might have been a continuation of the first error)

**Comment by emanuelschuetze** [comment]
Retested:
1. type error is fixed (confirmed)
2. I get no sortby arrow in sort menu afer opening motion list in a private tab (without indexedDB). Expected: arrow for identifier.

**Comment by FinnStutzenstein** [comment]
Rebased on the current master.

**Comment by tsiegleauq** [comment]
Rebased again, cleaned up, added better naming unified with current base-repo featuers.
Should work fine now, perhaps one or two setSortFunction calls are missing, but that should be pretty much it.

@emanuelschuetze please test

**Pull request merged**